dask/array/utils.py

```diff
@@ -565,6 +565,41 @@ def svd_flip(u, v, u_based_decision=False):
     return u, v
 
 
+def matmul_safe(a, b):
+    xp = np
+
+    if is_cupy_type(a):
+        import cupy
+
+        xp = cupy
+
+    chunk = xp.matmul(a, b)
+    # Since we have performed the contraction via matmul
+    # but blockwise expects all dimensions back, we need
+    # to add one dummy dimension back
+    return chunk[..., xp.newaxis]
+
+
+def scipy_linalg_safe(func_name, *args, **kwargs):
+    # need to evaluate at least the first input array
+    # for gpu/cpu checking
+    a = args[0]
+    if is_cupy_type(a):
+        import cupyx.scipy.linalg
+
+        func = getattr(cupyx.scipy.linalg, func_name)
+    else:
+        import scipy.linalg
+
+        func = getattr(scipy.linalg, func_name)
```

Comment:
I wonder if we could use `Dispatch` for this kind of thing

https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/sizeof.py#L17-L28

We could also then do lazy registration with this

https://github.com/dask/dask/blob/c5633c21e754df6b2c13195f1e2688e347c48963/dask/sizeof.py#L97-L106

Reply:
I think that's a good idea if this grows more.  As it stands now, this PR only adds `solve_triangular`.  I tried adding in `lu` but ran into a number of issues to due to lack of `int64` support in the cupy implementation

Reply:
FWIW I had some code here that might be relevant ( https://github.com/jakirkham/dask/commit/81adbc82c484436a33d4617eecf1b7f0eb84bf1b ). Though feel free to skip it if you would rather

Reply:
Also thinking about this in the context of dispatch refactoring work going on around Arrays and DataFrames ( https://github.com/dask/dask/pull/7505 ) ( https://github.com/dask/dask/pull/7503 )

Reply:
>  I tried adding in `lu` but ran into a number of issues to due to lack of `int64` support in the cupy implementation

Just to clarify is that suppose to be `float64`? While we may be able to provide `int64` arrays, expect these would likely need to be casted to `float64` and then the result would also be in `float64`. Though maybe I'm misunderstanding something here

Reply:
No, I think we need int64 support for `lu` functionality.  Or maybe a conversion is what scipy lu does and I'm unaware of it